File: src/lib/ordo.ts
import { useCallback, useEffect, useState } from "react";
import { BUILTIN_CATEGORIES, type Category, type CategoryId } from "./categories";

/**
 * Re-exported so the dozen modules that already import `CategoryId` from here
 * keep working now that the category list itself lives in `categories.ts`.
 */
export type { CategoryId };

/** How the app prints a wall-clock time. Chosen by the user, stored per account. */
export type HourFormat = "24h" | "12h";

export type Settings = {
  hourFormat: HourFormat;
};

export const DEFAULT_SETTINGS: Settings = { hourFormat: "24h" };

/**
 * First-run guess from the browser's own locale, so an en-US visitor sees
 * "9:00 AM" before touching anything. Falls back to 24h during SSR.
 */
export function detectHourFormat(): HourFormat {
  try {
    return new Intl.DateTimeFormat().resolvedOptions().hour12 ? "12h" : "24h";
  } catch {
    return "24h";
  }
}

/** Documents saved before settings existed have none — read through this. */
export const settingsOf = (state: OrdoState | null | undefined): Settings => ({
  ...DEFAULT_SETTINGS,
  ...(state?.settings ?? {}),
});

export const hourFormatOf = (state: OrdoState | null | undefined): HourFormat =>
  settingsOf(state).hourFormat;

/** `"14:05"` → `"14:05"` or `"2:05 PM"`. Garbage in comes straight back out. */
export function formatTime(hhmm: string, format: HourFormat = "24h"): string {
  const m = /^(\d{1,2}):(\d{2})/.exec(hhmm?.trim() ?? "");
  if (!m) return hhmm ?? "";
  const hours = Number(m[1]);
  const minutes = m[2]!;
  if (!Number.isFinite(hours) || hours > 23) return hhmm;
  if (format === "24h") return `${String(hours).padStart(2, "0")}:${minutes}`;
  const suffix = hours < 12 ? "AM" : "PM";
  const hour12 = hours % 12 === 0 ? 12 : hours % 12;
  return `${hour12}:${minutes} ${suffix}`;
}

/** The pair as one label: `"09:00–10:30"` or `"9:00 AM – 10:30 AM"`. */
export function formatTimeRange(start: string, end: string, format: HourFormat = "24h"): string {
  return format === "24h"
    ? `${formatTime(start, format)}–${formatTime(end, format)}`
    : `${formatTime(start, format)} – ${formatTime(end, format)}`;
}

export type Block = {
  id: string;
  title: string;
  start: string; // HH:MM
  end: string;
  category: CategoryId;
  priority: "must" | "nice";
  goalId?: string;
};

export type Goal = {
  id: string;
  title: string;
  period: "year" | "semester" | "month" | "week" | "day";
  category: CategoryId;
  target: number; // target completion %
  parentId?: string;
};

/** dateKey -> blockId -> percent complete (0-100) */
export type LogMap = Record<string, Record<string, number>>;

export type OrdoState = {
  /** 0 = Sunday .. 6 = Saturday */
  routine: Record<number, Block[]>;
  overrides: Record<string, Block[]>;
  templates: { id: string; name: string; blocks: Block[] }[];
  goals: Goal[];
  log: LogMap;
  journal: Record<string, string>;
  /** Absent in documents written before preferences existed. */
  settings?: Settings;
};

export const dateKey = (d: Date) =>
  `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;

export const addDays = (d: Date, n: number) => {
  const x = new Date(d);
  x.setDate(x.getDate() + n);
  return x;
};

export const startOfWeek = (d: Date) => addDays(d, -d.getDay());

const uid = () => Math.random().toString(36).slice(2, 10);

const b = (
  title: string,
  start: string,
  end: string,
  category: CategoryId,
  priority: "must" | "nice" = "must",
): Block => ({ id: uid(), title, start, end, category, priority });

const weekday: Block[] = [
  b("Fajr + reflection", "05:30", "06:00", "spiritual"),
  b("Workout", "06:00", "07:00", "health"),
  b("Deep work block", "09:00", "11:30", "work"),
  b("Study: algorithms", "14:00", "16:00", "study"),
  b("Read 30 pages", "21:00", "22:00", "study", "nice"),
];

const weekend: Block[] = [
  b("Long run", "07:00", "08:30", "health"),
  b("Weekly review", "10:00", "11:00", "work"),
  b("Family time", "16:00", "18:00", "relationships", "nice"),
  b("Budget check", "19:00", "19:30", "finance", "nice"),
];

const seedGoals: Goal[] = [
  {
    id: "g-year",
    title: "Become undeniably disciplined",
    period: "year",
    category: "work",
    target: 80,
  },
  {
    id: "g-sem",
    title: "Ship Ordo + pass semester",
    period: "semester",
    category: "study",
    target: 85,
    parentId: "g-year",
  },
  {
    id: "g-month",
    title: "Train 20 days this month",
    period: "month",
    category: "health",
    target: 75,
    parentId: "g-sem",
  },
  {
    id: "g-week",
    title: "10h focused study",
    period: "week",
    category: "study",
    target: 90,
    parentId: "g-month",
  },
];

function seedLog(routine: Record<number, Block[]>): LogMap {
  const log: LogMap = {};
  const today = new Date();
  for (let i = 1; i <= 70; i++) {
    const d = addDays(today, -i);
    const blocks = routine[d.getDay()] ?? [];
    const key = dateKey(d);
    log[key] = {};
    const bias = 0.45 + 0.4 * Math.sin(i / 9) + (d.getDay() === 0 ? -0.2 : 0);
    blocks.forEach((blk, idx) => {
      const r = Math.abs(Math.sin(i * 3.7 + idx * 1.3));
      log[key]![blk.id] = r < bias ? 100 : r < bias + 0.18 ? 50 : 0;
    });
  }
  return log;
}

export const defaultState = (): OrdoState => {
  const routine: Record<number, Block[]> = {
    0: weekend.map((x) => ({ ...x, id: uid() })),
    1: weekday.map((x) => ({ ...x, id: uid() })),
    2: weekday.map((x) => ({ ...x, id: uid() })),
    3: weekday.map((x) => ({ ...x, id: uid() })),
    4: weekday.map((x) => ({ ...x, id: uid() })),
    5: weekday.map((x) => ({ ...x, id: uid() })),
    6: weekend.map((x) => ({ ...x, id: uid() })),
  };
  return {
    routine,
    overrides: {},
    templates: [
      { id: uid(), name: "Exam week", blocks: weekday.map((x) => ({ ...x, id: uid() })) },
    ],
    goals: seedGoals,
    log: seedLog(routine),
    journal: {},
    settings: { ...DEFAULT_SETTINGS, hourFormat: detectHourFormat() },
  };
};

const KEY = "ordo.state.v1";

export function useOrdo() {
  const [state, setState] = useState<OrdoState | null>(null);

  useEffect(() => {
    try {
      const raw = localStorage.getItem(KEY);
      setState(raw ? (JSON.parse(raw) as OrdoState) : defaultState());
    } catch {
      setState(defaultState());
    }
  }, []);

  useEffect(() => {
    if (state) localStorage.setItem(KEY, JSON.stringify(state));
  }, [state]);

  const update = useCallback((fn: (s: OrdoState) => OrdoState) => {
    setState((s) => (s ? fn(s) : s));
  }, []);

  // Reset wipes the plan and the log, not the preferences — the clock format is
  // a display choice, not data the user asked to throw away.
  const reset = useCallback(() => {
    setState((prev) => ({ ...defaultState(), settings: settingsOf(prev) }));
  }, []);

  return { state, update, reset };
}

export function blocksFor(state: OrdoState, d: Date): Block[] {
  const key = dateKey(d);
  const list = state.overrides[key] ?? state.routine[d.getDay()] ?? [];
  return [...list].sort((a, x) => a.start.localeCompare(x.start));
}

export function dayScore(state: OrdoState, d: Date): number | null {
  const blocks = blocksFor(state, d);
  if (!blocks.length) return null;
  const entries = state.log[dateKey(d)] ?? {};
  const total = blocks.reduce((sum, blk) => sum + (entries[blk.id] ?? 0), 0);
  return Math.round(total / blocks.length);
}

export function rangeScore(state: OrdoState, from: Date, days: number): number {
  let sum = 0;
  let n = 0;
  for (let i = 0; i < days; i++) {
    const s = dayScore(state, addDays(from, i));
    if (s !== null) {
      sum += s;
      n++;
    }
  }
  return n ? Math.round(sum / n) : 0;
}

export function streak(state: OrdoState, threshold = 70): { current: number; best: number } {
  let current = 0;
  let best = 0;
  let run = 0;
  for (let i = 120; i >= 0; i--) {
    const s = dayScore(state, addDays(new Date(), -i));
    if (s !== null && s >= threshold) {
      run++;
      best = Math.max(best, run);
    } else if (s !== null) {
      run = 0;
    }
    if (i === 0 || run > 0) current = run;
  }
  return { current, best };
}

/**
 * Mean completion per category over `days`. The category list is passed in
 * because it is no longer a constant — an admin can add to it — and any id found
 * in the plan but missing from the list is still reported, under its raw id, so
 * a deleted category does not silently drop history.
 */
export function categoryBreakdown(
  state: OrdoState,
  days = 28,
  cats: Pick<Category, "id" | "label">[] = BUILTIN_CATEGORIES,
) {
  const acc: Record<string, { sum: number; n: number }> = {};
  for (let i = 0; i < days; i++) {
    const d = addDays(new Date(), -i);
    const entries = state.log[dateKey(d)] ?? {};
    for (const blk of blocksFor(state, d)) {
      const a = (acc[blk.category] ??= { sum: 0, n: 0 });
      a.sum += entries[blk.id] ?? 0;
      a.n += 1;
    }
  }
  const known = cats.map((c) => ({ id: c.id, label: c.label }));
  const extra = Object.keys(acc)
    .filter((id) => !known.some((c) => c.id === id))
    .map((id) => ({ id, label: id }));
  return [...known, ...extra].map((c) => ({
    category: c.label,
    id: c.id,
    value: acc[c.id]?.n ? Math.round(acc[c.id]!.sum / acc[c.id]!.n) : 0,
  }));
}

export function missedDebt(state: OrdoState) {
  const out: { date: string; block: Block }[] = [];
  for (let i = 1; i <= 14; i++) {
    const d = addDays(new Date(), -i);
    const entries = state.log[dateKey(d)] ?? {};
    for (const blk of blocksFor(state, d)) {
      if (blk.priority === "must" && (entries[blk.id] ?? 0) < 50) {
        out.push({ date: dateKey(d), block: blk });
      }
    }
  }
  return out.slice(0, 8);
}

export const newBlock = (): Block => ({
  id: uid(),
  title: "New block",
  start: "08:00",
  end: "09:00",
  category: "work",
  priority: "must",
});

export const cloneBlocks = (blocks: Block[]) => blocks.map((x) => ({ ...x, id: uid() }));
export const newId = uid;

/**
 * The clock preference without mounting the document hooks — for pages like the
 * admin console that print times but have no business loading someone's plan.
 * Falls back to the browser locale during SSR or before a first save.
 */
export function storedHourFormat(): HourFormat {
  try {
    const raw = localStorage.getItem(KEY);
    if (raw) return hourFormatOf(JSON.parse(raw) as OrdoState);
  } catch {
    /* unreadable storage is not worth an error */
  }
  return detectHourFormat();
}

File: mobile/lib/utils/ordo.dart
import '../models/ordo_state.dart';

String dateKey(DateTime d) =>
    '${d.year}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';

int dayOfWeek(DateTime d) => d.weekday % 7;

DateTime addDays(DateTime d, int n) => DateTime(d.year, d.month, d.day + n);

DateTime startOfWeek(DateTime d) => addDays(d, -(d.weekday % 7));

String formatTime(String hhmm, {bool hour12 = false}) {
  final m = RegExp(r'^(\d{1,2}):(\d{2})').firstMatch(hhmm);
  if (m == null) return hhmm;
  final hours = int.parse(m.group(1)!);
  final minutes = m.group(2)!;
  if (!hour12) return '${hours.toString().padLeft(2, '0')}:$minutes';
  final suffix = hours < 12 ? 'AM' : 'PM';
  final h12 = hours % 12 == 0 ? 12 : hours % 12;
  return '$h12:$minutes $suffix';
}

String formatTimeRange(String start, String end, {bool hour12 = false}) {
  if (hour12) {
    return '${formatTime(start, hour12: true)} – ${formatTime(end, hour12: true)}';
  }
  return '${formatTime(start)}–${formatTime(end)}';
}

List<Block> blocksFor(OrdoState state, DateTime d) {
  final key = dateKey(d);
  final list = state.overrides[key] ?? state.routine[dayOfWeek(d)] ?? [];
  return List<Block>.from(list)..sort((a, b) => a.start.compareTo(b.start));
}

int dayScore(OrdoState state, DateTime d) {
  final blocks = blocksFor(state, d);
  if (blocks.isEmpty) return -1;
  final entries = state.log[dateKey(d)] ?? {};
  final total = blocks.fold(0, (sum, blk) => sum + (entries[blk.id] ?? 0));
  return (total / blocks.length).round();
}

int rangeScore(OrdoState state, DateTime from, int days) {
  var sum = 0;
  var n = 0;
  for (var i = 0; i < days; i++) {
    final s = dayScore(state, addDays(from, i));
    if (s >= 0) {
      sum += s;
      n++;
    }
  }
  return n > 0 ? (sum / n).round() : 0;
}

Streak streak(OrdoState state) {
  var current = 0;
  var best = 0;
  var run = 0;
  for (var i = 120; i >= 0; i--) {
    final s = dayScore(state, addDays(DateTime.now(), -i));
    if (s >= 70) {
      run++;
      if (run > best) best = run;
    } else if (s >= 0) {
      run = 0;
    }
    if (i == 0 || run > 0) current = run;
  }
  return Streak(current: current, best: best);
}

Map<String, dynamic> categoryBreakdown(OrdoState state, List<Category> cats, {int days = 28}) {
  final acc = <String, Map<String, double>>{};
  for (var i = 0; i < days; i++) {
    final d = addDays(DateTime.now(), -i);
    final key = dateKey(d);
    final entries = state.log[key] ?? {};
    for (final blk in blocksFor(state, d)) {
      if (!acc.containsKey(blk.category)) {
        acc[blk.category] = {'sum': 0, 'n': 0};
      }
      acc[blk.category]!['sum'] = (acc[blk.category]!['sum'] ?? 0) + (entries[blk.id] ?? 0);
      acc[blk.category]!['n'] = (acc[blk.category]!['n'] ?? 0) + 1;
    }
  }
  final known = cats.map((c) => {'id': c.id, 'label': c.label}).toList();
  final extra = acc.keys.where((id) => !cats.any((c) => c.id == id)).map((id) => {'id': id, 'label': id}).toList();
  return {
    'cats': [...known, ...extra].map((c) => {
      'category': c['label'],
      'id': c['id'],
      'value': (acc[c['id']]?['n'] ?? 0) > 0 ? (((acc[c['id']]?['sum'] ?? 0) / (acc[c['id']]?['n'] ?? 1)).round()) : 0,
    }).toList(),
  };
}

List<Map<String, dynamic>> missedDebt(OrdoState state) {
  final out = <Map<String, dynamic>>[];
  for (var i = 1; i <= 14; i++) {
    final d = addDays(DateTime.now(), -i);
    final entries = state.log[dateKey(d)] ?? {};
    for (final blk in blocksFor(state, d)) {
      if (blk.priority == 'must' && (entries[blk.id] ?? 0) < 50) {
        out.add({'date': dateKey(d), 'block': blk});
      }
    }
  }
  return out.take(8).toList();
}

Block newBlock() => Block(
  id: DateTime.now().millisecondsSinceEpoch.toString(),
  title: 'New block',
  start: '08:00',
  end: '09:00',
  category: 'work',
  priority: 'must',
);

List<Block> cloneBlocks(List<Block> blocks) => blocks.map((b) => b.copyWith(id: DateTime.now().millisecondsSinceEpoch.toString())).toList();

Streak streakFromState(OrdoState state) => streak(state);

class Streak {
  final int current;
  final int best;
  const Streak({required this.current, required this.best});
}

class HourFormat {
  static const twentyFour = '24h';
  static const twelve = '12h';
}

String hourFormatOf(OrdoState? state) => state?.settings?.hourFormat ?? '24h';

bool isHour12(OrdoState? state) => hourFormatOf(state) == '12h';

List<Map<String, dynamic>> weeklyCompletion(OrdoState state) {
  final out = <Map<String, dynamic>>[];
  for (var w = 9; w >= 0; w--) {
    final from = startOfWeek(addDays(DateTime.now(), -7 * w));
    out.add({
      'label': '${from.month}/${from.day}',
      'value': rangeScore(state, from, 7).toDouble(),
    });
  }
  return out;
}

List<Map<String, dynamic>> categoryBalance(OrdoState state, List<Category> cats) {
  final cats2 = categoryBreakdown(state, cats);
  return (cats2['cats'] as List).map((e) => {'category': e['category'], 'id': e['id'], 'value': (e['value'] as num).toDouble()}).toList();
}

Map<String, dynamic> yearInReview(OrdoState state, List<Category> cats) {
  final days = 365;
  var sum = 0.0;
  var n = 0;
  var bestWeek = 0;
  for (var i = days - 1; i >= 0; i--) {
    final s = dayScore(state, addDays(DateTime.now(), -i));
    if (s >= 0) {
      sum += s;
      n++;
    }
  }
  for (var w = 51; w >= 0; w--) {
    final ws = rangeScore(state, startOfWeek(addDays(DateTime.now(), -7 * w)), 7);
    if (ws > bestWeek) bestWeek = ws;
  }
  final cats2 = categoryBreakdown(state, cats, days: 365);
  final catList = (cats2['cats'] as List).cast<Map<String, dynamic>>();
  String? mostConsistentLabel;
  if (catList.isNotEmpty) {
    var bestVal = -1;
    for (final c in catList) {
      final v = (c['value'] as num).toInt();
      if (v > bestVal) {
        bestVal = v;
        mostConsistentLabel = c['category'] as String?;
      }
    }
  }
  final avg = n > 0 ? (sum / n).round() : 0;
  return {
    'avg': avg,
    'bestWeek': bestWeek,
    'mostConsistent': mostConsistentLabel ?? '—',
    'daysTracked': n,
    'isNew': n < 7,
    'avgDouble': n > 0 ? (sum / n) : 0.0,
  };
}

Map<String, dynamic> badges(Streak s, int month) {
  final earned = <String>[];
  if (s.best >= 7) earned.add('7-day streak');
  if (s.best >= 30) earned.add('30-day streak');
  if (s.best >= 60) earned.add('60-day streak');
  if (month >= 70) earned.add('70% month');
  if (s.current >= 3 && month < 40) earned.add('Comeback');
  return {'earned': earned, 'current': s.current, 'best': s.best};
}
